### PR TITLE
[Feature] Support sequence packing

### PR DESCRIPTION
**Pull request by yytang0204**
Feature:
- Support microbatch regroup within minibatch
- Support packing for train and forward
- Add more args check
- Change loss aggr type to average on token-level loss within sample, average across samples

Tested on qwen3-8B and qwen3-30B-A3B
image

### CODE REVIEW

File: chatlearn/algorithm/grpo_utils/advantage_compute.py [hunk 1]
```diff
@@ -20,4 +20,6 @@ def compute_grpo_adv(episode_replay_buffers):
         for i, li in enumerate(l):
             li["advantages"] = (rewards[i] - mean) / (std + 1e-5)
         res_buffers.extend(l)
+    # Shuffle train sample to balance training workload across dp_ranks
+    random.shuffle(res_buffers)
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/advantage_compute.py:24]: 有具体实验数据么，会不会导致mcore那边计算ref和action logprob时候产生数据乱序

File: chatlearn/algorithm/grpo_utils/trainer_utils.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    attn_mask = torch.zeros_like(tokens, dtype=torch.int32, device=tokens.device)
+    for i in range(len(prompt_token_length)):
+        loss_mask[i, prompt_token_length[i]: prompt_token_length[i] + response_token_length[i]] = 1.0
+        attn_mask[i, : prompt_token_length[i] + response_token_length[i]] = 1.0
+    _, seq_len = tokens.size()
+    position_ids = torch.arange(seq_len, dtype=torch.long, device=tokens.device)
+    position_ids = position_ids.unsqueeze(0).expand_as(tokens)
+
+    return attn_mask, loss_mask, position_ids
+def bin_packing(seq_len_list, max_train_token):
```
Comment (haiasd) [inline chatlearn/algorithm/grpo_utils/trainer_utils.py:38]: 可以给下面几个方法加上类型注释么

**Review by jerryli1981 — APPROVED** [review]

File: chatlearn/algorithm/grpo_utils/policy_trainer.py [hunk 3]
```diff
@@ -5,204 +5,165 @@
...
-            "old_logprobs": old_logprobs,
-            "ref_logprobs": ref_logprobs,
-        }
-
-        for k, v in inputs.items():
-            inputs[k] = to_device(torch.cuda.current_device(), v)
-
-        return inputs
+    def preprocess_data_list(self, data_list, training:bool):
+        top = time.time()
```
[the remark this answers is not in the document]
Comment (yytang0204) [reply chatlearn/algorithm/grpo_utils/policy_trainer.py]: removed
